### DISCUSSION TITLE
Add --dump-hw-params option (alsa backend)

### DISCUSSION THREAD

**Discussion by JasonLG1979** [librespot-org/librespot#799] [Feature Requests]
This function in the alsa backend (along with the other bits necessary) could add the option to dump the hardware parameters of a device mimicking the output of `aplay`'s `--dump-hw-params` option.

```rust
// needs Output from alsa
// use alsa::{Direction, Error, Output, ValueOr};

...

fn dump_hw_params(dev_name: &str) {
    let pcm = PCM::new(dev_name, Direction::Playback, false).unwrap();
    let hwp = HwParams::any(&pcm).unwrap();
    let mut output = Output::buffer_open().unwrap();
    hwp.dump(&mut output).unwrap();

    println!("\nHW Params of device {}\n", dev_name);
    println!("--------------------\n");
    println!("{}", output);
    println!("--------------------\n");
    exit(0);
}
```

The function would take the same arguments as `--device` so for example a user could do `librespot --dump-hw-params hw:2,0` and they would get something like this:
```
HW Params of device hw:2,0

--------------------

ACCESS:  MMAP_INTERLEAVED RW_INTERLEAVED
FORMAT:  S16_LE S24_3LE
SUBFORMAT:  STD
SAMPLE_BITS: [16 24]
FRAME_BITS: [32 48]
CHANNELS: 2
RATE: [32000 96000]
PERIOD_TIME: [1000 4096000]
PERIOD_SIZE: [32 131072]
PERIOD_BYTES: [128 524288]
PERIODS: [2 1024]
BUFFER_TIME: (666 8192000]
BUFFER_SIZE: [64 262144]
BUFFER_BYTES: [256 1048576]
TICK_TIME: ALL

--------------------
```

I think it would be super helpful for debugging configuring.

**Comment by kingosticks** [comment]
My opinion is to add this info to the (debug) logging - if it's not already there- but this is not necessary as an extra dedicated command line argument.

**Reply by JasonLG1979** [reply]
It's not already there. It could easily be added that way I suppose, but the idea for an arg would be it would make it super easy for a user to just call `librespot --dump-hw-params hw:x,x`  to see what formats and sampling rates their device supported. Combined with the existing `--device ?` arg to list output devices it would make getting info to configure `librespot` super easy.

**Reply by JasonLG1979** [reply]
I see dumping the hw-params after we've set the hw-params more of a useful debug output. Dumping the hw-params before they're configured (like above) gives you the capabilities of the hardware. Dumping then after lets you verify things are configured as intended.

This is the output after  we set the hw-params:

```
HW Params of device hw:2,0

--------------------

ACCESS:  RW_INTERLEAVED
FORMAT:  S24_3LE
SUBFORMAT:  STD
SAMPLE_BITS: 24
FRAME_BITS: 48
CHANNELS: 2
RATE: 44100
PERIOD_TIME: (125011 125012)
PERIOD_SIZE: 5513
PERIOD_BYTES: 33078
PERIODS: 4
BUFFER_TIME: (500045 500046)
BUFFER_SIZE: 22052
BUFFER_BYTES: 132312
TICK_TIME: 0

--------------------
```

**Reply by roderickvd** [reply]
Not sure. Pre-setup, I don't see why we would want to copy `aplay` which will be present on all systems that have Alsa anyway. Post-setup, the Alsa bindings are such that they will fail (panic) if the requested setup could not be configured, so why verify?

Also agree with @kingosticks that regardless, I don't think it warrants a new command line option, especially not when it's Alsa-only.

**Reply by JasonLG1979** [reply]
> Not sure. Pre-setup, I don't see why we would want to copy `aplay` which will be present on all systems that have Alsa anyway.

To get the info out of `aplay` you actually have to play audio so you'd need to do something like `aplay -Dhw:0,0 --dump-hw-params foo.wav` You can't just do `aplay -Dhw:0,0 --dump-hw-params`.

> Post-setup, the Alsa bindings are such that they will fail (panic) if the requested setup could not be configured, so why verify?

I agree that it's not very useful in the debug output I was just saying it would be more use post than pre.

**Reply by JasonLG1979** [reply]
Plus I think there's a little bit of a precedent to following `aplay`'s lead to get hardware info. `--device ?` outputs the same thing as `aplay -L`.

**Reply by JasonLG1979** [reply]
> the Alsa bindings are such that they will fail (panic) if the requested setup could not be configured

That's not exactly true. Every time you see a ` ValueOr::Nearest` that means you're asking alsa not telling it. They may not fail but that doesn't mean you actually get what you want. But really you can get all the post config info you need from `cat /proc/asound/card*/pcm*p/sub*/hw_params`.

**Reply by roderickvd** [reply]
> To get the info out of `aplay` you actually have to play audio so you'd need to do something like `aplay -Dhw:0,0 --dump-hw-params foo.wav` You can't just do `aplay -Dhw:0,0 --dump-hw-params`.

Isn't that what everyone does with the WAVs in `/usr/share/sounds/alsa`?
 
> That's not exactly true. Every time you see a ` ValueOr::Nearest` that means you're asking alsa not telling it. They may not fail but that doesn't mean you actually get what you want. But really you can get all the post config info you need from `cat /proc/asound/card*/pcm*p/sub*/hw_params`.

That's right I use this little script to do exactly that (not mine):

```bash
#!/bin/bash

if [ "$1" = '-h' -o "$1" = '--help' ]; then
        me="${0##*/}"
        echo "$me: list ALSA devices and relevant information."
        exit 0
fi

if [ -e '/proc/asound/version' ]; then
        cat '/proc/asound/version'
        echo
fi

for (( i=0; i<100; i++ )); do
        if [ -d "/proc/asound/card${i}" ]; then
                cname="$( cat "/proc/asound/card${i}/id" )"
                echo "Card ${i} (${cname}):"

                for t in 'p' 'c'; do
                        case "$t" in
                                p) typeString='Playback' ;;
                                c) typeString='Recording' ;;
                        esac

                        for (( j=0; j<100; j++ )); do
                                if [ -d "/proc/asound/card${i}/pcm${j}${t}" ]; then
                                        dname=''
                                        if [ -e "/proc/asound/card${i}/pcm${j}${t}/info" ]; then
                                                dname="$( grep -E '^name: ' "/proc/asound/card${i}/pcm${j}${t}/info" 2>/dev/null | cut -d '                                                   ' -f 2- )"
                                        fi
                                        echo
                                        if [ -n "$dname" ]; then
                                                echo "  * ${typeString} Device ${j} (${dname}):"
                                        else
                                                echo "  * ${typeString} Device ${j}:"
                                        fi
                                        ownerPID="$( grep -E '^owner_pid' "/proc/asound/card${i}/pcm${j}${t}/sub0/status" 2>/dev/null | tr                                                   -cd '0-9' )"
                                        if [ -n "$ownerPID" ]; then
                                                tgid="$( grep -F 'Tgid:' "/proc/${ownerPID}/status" 2>/dev/null | tr -cd '0-9' )"
                                                if [ -n "$tgid" ]; then
                                                        ownerPID="$tgid"
                                                fi
                                                ownerName="$( cat "/proc/${ownerPID}/comm" 2>/dev/null )"
                                                echo "    used by: ${ownerName} (PID ${ownerPID})"
                                        fi
                                        while read line; do
                                                echo "    $line"
                                        done < "/proc/asound/card${i}/pcm${j}${t}/sub0/hw_params"
                                fi
                        done
                done
                echo
        fi
done
```

**Reply by JasonLG1979** [reply]
> Isn't that what everyone does with the WAVs in /usr/share/sounds/alsa?

Yes, `aplay -Dhw:x,x --dump-hw-params /usr/share/sounds/alsa/Front_Right.wav` gets you the same thing as above provided `/usr/share/sounds/alsa/Front_Right.wav` exists, although I've never know it not to?

I was thinking more about a non-Linux user who's trying to set up `librespot` for the 1st time on a Raspberry Pi or whatever. We forget pretty fast how steep the learning curve can be and how intimidating and disconcerting a headless system can be if you're coming from a whole life of GUI's. I was thinking basically let's make `librespot` a one stop shop. Let's make it so you can gather pretty much all the info you need to configure `librespot` with `librespot`.

> That's right I use this little script to do exactly that (not mine)

`cat /proc/asound/card*/pcm*p/sub*/hw_params` gets you everything except `* Playback Device X:` and `used by: XXXXX (PID xxxx)`

**Reply by JasonLG1979** [reply]
Honestly I'm not sure why you can't just do `aplay -Dhw:x,x --dump-hw-params` with `aplay` it seems kinda stupid that it fails. It's not like something *has* to be played to dump the hw params?

**Reply by roderickvd** [reply]
> > Isn't that what everyone does with the WAVs in /usr/share/sounds/alsa?
> 
> Yes, `aplay -Dhw:x,x --dump-hw-params /usr/share/sounds/alsa/Front_Right.wav` gets you the same thing as above provided `/usr/share/sounds/alsa/Front_Right.wav` exists, although I've never know it not to?

I bet distributions package it together with `aplay`. 
 
> I was thinking more about a non-Linux user who's trying to set up `librespot` for the 1st time on a Raspberry Pi or whatever. We forget pretty fast how steep the learning curve can be and how intimidating and disconcerting a headless system can be if you're coming from a whole life of GUI's. I was thinking basically let's make `librespot` a one stop shop. Let's make it so you can gather pretty much all the info you need to configure `librespot` with `librespot`.

Couple of thoughts:

1. Out-of-the-box we already offer a conservative, should-just-work solution. 16-bit PCM fed to Rodio.
2. For Alsa, very few devices that won't take S16 either.
3. Downstream packages and distributions can (and do) take further care to integrate `librespot` in a user-friendly manner.
4. Personally I am more than fine with users reporting configuration issues with `raspotify`, `ncspot`, moOde, and the like. You might have noticed how I'm moving for a fast close on such issues reported with us. If it's a bug, we should fix it, if it's a configuration issue, then that's where it's at.  I'd like to spend my time improving `librespot`, not guiding novice Linux users. Not to say that's not important -- we all had to learn at some point -- but not my personal goal nor do I think the goal of this project. Beginners can opt for an easy-to-use, repackaged version and go from there.

I don't want to make a bikeshed out of this. I mean if it can be integrated in an easy way without introducing new or Alsa-specific command line options, then I don't see why not. But piggying back on your idea to make `librespot` an integrated package, I say that's what those downstream integrated packages are for.
 
> Honestly I'm not sure why you can't just do `aplay -Dhw:x,x --dump-hw-params` with `aplay` it seems kinda stupid that it fails. It's not like something _has_ to be played to dump the hw params?

I've just about stopped wondering about Alsa idiosyncrasies. 😜

**Reply by JasonLG1979** [reply]
Far enough. It's not an absolute must. Just a thought. I don't want to make it a bikeshed either. It's not worth it. It does kinda remind me of the old adage that given enough time and feature creep that all applications will eventually be able to send emails,lol!!!

I appreciate the fact that this late in the 0.3 cycle it's really about fixing what's there and not adding new features. I'm a big fan of at some point in the cycle saying "Ok, no more features, let's thoroughly look at and test what we have for the rest of the cycle and only push fixes".

**Reply by JasonLG1979** [reply]
Since we're past 0.3 I was wondering if we could revisit this @roderickvd? My thought is maybe add it to `--device ?` so it could look something like this:

```
Listing available Alsa outputs:

----------------------------------------

Device:

hw:CARD=DAC,DEV=0

JDS Labs Element DAC, USB Audio
Direct hardware device without any conversions

HW Params:

ACCESS:  MMAP_INTERLEAVED RW_INTERLEAVED
FORMAT:  S16_LE S24_3LE
SUBFORMAT:  STD
SAMPLE_BITS: [16 24]
FRAME_BITS: [32 48]
CHANNELS: 2
RATE: [32000 96000]
PERIOD_TIME: [1000 4096000]
PERIOD_SIZE: [32 131072]
PERIOD_BYTES: [128 524288]
PERIODS: [2 1024]
BUFFER_TIME: (666 8192000]
BUFFER_SIZE: [64 262144]
BUFFER_BYTES: [256 1048576]
TICK_TIME: ALL

----------------------------------------

```

So it would be like a combination of what we already have (which mimics `aplay -L`) and `--dump-hw-params` and doesn't add any new alsa specific options. It just expands on what we already have.

**Reply by JasonLG1979** [reply]
Or an even better option would be to filter out devices that are not compatible and show what `librespot` formats the device supports? So it would show something like this:
```
Compatible alsa devices:

----------------------------------------

Device:

hw:CARD=DAC,DEV=0

Description:

JDS Labs Element DAC, USB Audio
Direct hardware device without any conversions

Supported Formats:

S16, S24_3

----------------------------------------

Device:

hw:CARD=DAC,DEV=1

Description:

JDS Labs Element DAC, USB Audio #1
Direct hardware device without any conversions

Supported Formats:

S16

----------------------------------------

Device:

plughw:CARD=DAC,DEV=0

Description:

JDS Labs Element DAC, USB Audio
Hardware device with all software conversions

Supported Formats:

S16, S24, S24_3, S24_3, S32, F32, F64

----------------------------------------

Device:

plughw:CARD=DAC,DEV=1

Description:

JDS Labs Element DAC, USB Audio #1
Hardware device with all software conversions

Supported Formats:

S16, S24, S24_3, S24_3, S32, F32, F64

----------------------------------------

Device:

sysdefault:CARD=DAC

Description:

JDS Labs Element DAC, USB Audio
Default Audio Device

Supported Formats:

S16, S24, S24_3, S24_3, S32, F32, F64

----------------------------------------

```

I actually like this better than trying to emulate ALSA. It only gives the user useful info. It would not show devices that can't do 2 ch 44.1 Interleaved.

```rust
fn list_outputs() -> SinkResult<()> {
    println!("\nCompatible alsa devices:\n");
    println!("----------------------------------------\n");

    let i = HintIter::new_str(None, "pcm").map_err(|_| AlsaError::Parsing)?;

    for a in i {
        if let Some(Direction::Playback) = a.direction {
            let name = a.name.ok_or(AlsaError::Parsing)?;
            let desc = a.desc.ok_or(AlsaError::Parsing)?;

            if let Ok(pcm) = PCM::new(&name, Direction::Playback, false) {
                if let Ok(hwp) = HwParams::any(&pcm) {
                    // Only show devices that support
                    // 2 ch 44.1 Interleaved.
                    if hwp.test_rate(SAMPLE_RATE).is_ok()
                            && hwp.test_channels(NUM_CHANNELS as u32).is_ok()
                            // There is no `test_access` in the bindings.
                            && hwp.set_access(Access::RWInterleaved).is_ok()
                    {
                        println!("Device:\n\n{}\n", name);
                        println!("Description:\n\n{}\n", desc);

                        let mut supported_formats = Vec::new();

                        for f in [
                            Format::s16(),
                            Format::s24(),
                            Format::S243LE,
                            Format::S243BE,
                            Format::s32(),
                            Format::float(),
                            Format::float64(),
                        ] {
                            if hwp.test_format(f).is_ok() {
                                let f = match f {
                                    Format::Float64LE | Format::Float64BE => AudioFormat::F64,
                                    Format::FloatLE | Format::FloatBE => AudioFormat::F32,
                                    Format::S32LE | Format::S32BE => AudioFormat::S32,
                                    Format::S24LE | Format::S24BE => AudioFormat::S24,
                                    Format::S243LE | Format::S243BE => AudioFormat::S24_3,
                                    Format::S16LE | Format::S16BE => AudioFormat::S16,
                                    _ => unreachable!(),
                                };

                                supported_formats.push(format!("{:?}", f));
                            }
                        }
                        println!("Supported Formats:\n\n{}\n", supported_formats.join(", "));
                        println!("----------------------------------------\n");
                    }
                };
            }
        }
    }

    Ok(())
}
```
```

**Reply by roderickvd** [reply]
If I understand correctly, you are suggesting to enhance the existing command-line `--device ?` option and have it report capabilities that are useful to `librespot`. I agree that that'd be something nice to have.

**Reply by JasonLG1979** [reply]
Yes. Basically instead of showing the user a bunch of devices that may not be compatible and a bunch of extraneous info, just show useful info.

**Reply by roderickvd** [reply]
OK let's continue at #883.

**Reply by JasonLG1979** [reply]
OK
